registry/storage/driver/gcs/gcs.go

```diff
@@ -336,64 +363,63 @@ func retry(maxTries int, req request) error {
 	return err
 }
 
-func (d *driver) writeCompletely(context ctx.Context, path string, offset int64, reader io.Reader) (totalRead int64, err error) {
-	wc := storage.NewWriter(d.context(context), d.bucket, d.pathToKey(path))
-	wc.ContentType = "application/octet-stream"
-	defer wc.Close()
-
-	// Copy the first offset bytes of the existing contents
-	// (padded with zeros if needed) into the writer
-	if offset > 0 {
-		existing, err := d.ReadStream(context, path, 0)
-		if err != nil {
-			return 0, err
-		}
-		defer existing.Close()
-		n, err := io.CopyN(wc, existing, offset)
-		if err == io.EOF {
-			err = writeZeros(wc, offset-n)
-		}
-		if err != nil {
-			return 0, err
-		}
+func (d *driver) getUploadSession(context ctx.Context, path string) (*uploadSession, error) {
+	res, err := d.getObject(context, path, 0)
+	if err != nil {
+		return nil, err
 	}
-	return io.Copy(wc, reader)
-}
-
-func skip(reader io.Reader, count int64) (int64, error) {
-	if count <= 0 {
-		return 0, nil
+	defer res.Body.Close()
+	if res.Header.Get("Content-Type") != uploadSessionContentType {
+		return nil, storagedriver.PathNotFoundError{Path: path}
+	}
+	offset, err := strconv.ParseInt(res.Header.Get("X-Goog-Meta-Offset"), 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	buffer, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
 	}
-	return io.CopyN(ioutil.Discard, reader, count)
+	return &uploadSession{
+		name:       d.pathToKey(path),
+		sessionURI: res.Header.Get("X-Goog-Meta-Session-URI"),
+		buffer:     buffer,
+		offset:     offset,
+	}, nil
 }
 
-func writeZeros(wc io.Writer, count int64) error {
-	buf := make([]byte, 32*1024)
-	for count > 0 {
-		size := cap(buf)
-		if int64(size) > count {
-			size = int(count)
-		}
-		n, err := wc.Write(buf[0:size])
-		if err != nil {
-			return err
-		}
-		count = count - int64(n)
+func (d *driver) updateUploadSession(context ctx.Context, s *uploadSession) error {
+	wc := storage.NewWriter(d.context(context), d.bucket, s.name)
+	wc.ContentType = uploadSessionContentType
+	wc.Metadata = map[string]string{
+		"Session-URI": s.sessionURI,
+		"Offset":      fmt.Sprintf("%v", s.offset),
```

Comment:
This could be `Sprint(s.offset)`, or better yet, `strconv.FormatInt(s.offset, 10)`. It's just nice to avoid format strings and reflection when possible.